File: highway-pursuit-server/highway-pursuit-server/Injected/CheatService.cs
using HighwayPursuitServer.Data;
using HighwayPursuitServer.Server;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.InteropServices;
using System.Text;
using System.Threading.Tasks;

namespace HighwayPursuitServer.Injected
{
    class CheatService
    {
        private readonly IHookManager _hookManager;

        public CheatService(IHookManager hookManager)
        {
            this._hookManager = hookManager;
            this.RegisterHooks();
        }

        #region Hooking
        private void RegisterHooks()
        {
            IntPtr getLivesPtr = new IntPtr(_hookManager.GetModuleBase().ToInt32() + MemoryAdresses.GET_LIVES_OFFSET);
            GetLives = Marshal.GetDelegateForFunctionPointer<GetLives_delegate>(getLivesPtr);
            _hookManager.RegisterHook(getLivesPtr, new GetLives_delegate(GetLives_Hook));
        }

        #region hooks
        byte GetLives_Hook()
        {
            return HighwayPursuitConstants.CHEATED_CONSTANT_LIVES;
        }
        #endregion

        #region delegates
        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
        [return: MarshalAs(UnmanagedType.I1)]
        delegate byte GetLives_delegate();
        #endregion

        #region original function pointers
        static GetLives_delegate GetLives;
        #endregion
        #endregion
    }
}

File: highway-pursuit-server/highway-pursuit-server/Injected/CheatService.cpp
#include "../pch.h"
#include "CheatService.hpp"
#include "MemoryAddresses.hpp"

namespace Injected
{
    // Constructor
    CheatService::CheatService(std::shared_ptr<HookManager> hookManager)
        : _hookManager(hookManager)
    {
        RegisterHooks();
    }

    // RegisterHooks method
    void CheatService::RegisterHooks()
    {
        CheatService::Instance = this;

        LPVOID getLivesPtr = reinterpret_cast<LPVOID>(_hookManager->GetModuleBase() + MemoryAddresses::GET_LIVES_OFFSET);
        _hookManager->RegisterHook(getLivesPtr, &GetLives_StaticHook, &GetLive_Base);
    }

    uint32_t CheatService::GetLives_Hook()
    {
        return Data::HighwayPursuitConstants::CHEATED_CONSTANT_LIVES;
    }

    uint32_t __cdecl CheatService::GetLives_StaticHook(void)
    {
        if (CheatService::Instance != nullptr)
        {
            return CheatService::Instance->GetLives_Hook();
        }
        return Data::HighwayPursuitConstants::CHEATED_CONSTANT_LIVES;
    }

    CheatService* CheatService::Instance = nullptr;
    CheatService::GetLives_t CheatService::GetLive_Base = nullptr;
}
